packages/flutter/lib/src/material/tabs.dart

```diff
@@ -760,543 +352,478 @@ class TabBar<T> extends Scrollable implements AppBarBottomWidget {
   /// the color of the theme's body2 text color is used.
   final Color labelColor;
 
-  /// The height of the tab labels and indicator.
   @override
   double get bottomHeight {
-    for (TabLabel label in labels.values) {
-      if (label.hasText && label.hasIcon)
-        return _kTextAndIconTabHeight + _kTabIndicatorHeight;
+    for (Widget widget in tabs) {
+      if (widget is Tab) {
+        final Tab tab = widget;
+        if (tab.text != null && tab.icon != null)
+          return _kTextAndIconTabHeight + _kTabIndicatorHeight;
+      }
     }
     return _kTabHeight + _kTabIndicatorHeight;
   }
 
   @override
-  _TabBarState<T> createState() => new _TabBarState<T>();
+  _TabBarState createState() => new _TabBarState();
 }
 
-class _TabBarState<T> extends ScrollableState<TabBar<T>> implements TabBarSelectionAnimationListener {
-  TabBarSelectionState<T> _selection;
-  bool _valueIsChanging = false;
-  int _lastSelectedIndex = -1;
+class _TabBarState extends State<TabBar> {
+  final GlobalKey<ScrollableState> viewportKey = new GlobalKey<ScrollableState>();
 
-  void _initSelection(TabBarSelectionState<T> newSelection) {
-    if (_selection == newSelection)
-      return;
-    _selection?.removeAnimationListener(this);
-    _selection = newSelection;
-    _selection?.addAnimationListener(this);
-    if (_selection != null)
-      _lastSelectedIndex = _selection.index;
-  }
+  TabController _controller;
+  _ChangeAnimation _changeAnimation;
+  _IndicatorPainter _indicatorPainter;
+  int _currentIndex;
 
-  @override
-  void didUpdateConfig(TabBar<T> oldConfig) {
-    super.didUpdateConfig(oldConfig);
-    if (config.isScrollable != oldConfig.isScrollable) {
-      scrollBehavior.isScrollable = config.isScrollable;
-      if (!config.isScrollable)
-        scrollTo(0.0);
+  void _initTabController() {
+    if (_controller != null)
+      _controller.animation.removeListener(_handleTick);
+    _controller = config.controller ?? DefaultTabController.of(context);
```

Comment:
You might want to check whether the new tab controller is the same as the old tab controller before going through all this work to rewire the listeners and such.  For example, the dependencies might have changed for some other reason (e.g., the them changing or something).